aten/src/ATen/native/nested/NestedTensorMath.cpp

```diff
@@ -60,15 +60,23 @@ Tensor pad_tensor_to_shape(
 
 
 Tensor NestedTensor_nested_tensor_from_mask(const Tensor& t, const Tensor& mask, bool mask_check) {
-    TORCH_CHECK(mask.scalar_type() == at::ScalarType::Bool, "Expected mask to be of ScalarType Bool, but got ", mask.scalar_type(), " instead.");
+  TORCH_CHECK((mask.scalar_type() == at::ScalarType::Bool) || (at::is_floating_point(mask)),
+                "Expected mask to be of type Bool or floating dtype but got ",
+                mask.scalar_type(), " instead.");
     TORCH_CHECK(mask.dim() == 2, "Padding mask should be 2D");
     TORCH_CHECK(t.dim() == 3, "Input should be a 3D tensor, N * L * D");
     auto N = t.size(0), L = t.size(1), D = t.size(2);
     auto NN = mask.size(0), LL = mask.size(1);
     TORCH_CHECK(N == NN && L == LL, "Mask size should match input size");
 
     // N * L
-    Tensor sizes = mask;
+    Tensor sizes;
+    if (mask.scalar_type() != at::ScalarType::Bool){
+      sizes = mask.to(at::kBool);
```

Comment:
Is my understanding here correct? If the user passes a bool mask into `TransformerEncoderLayer`, it will be converted to a float in `F._canonical_mask()`. Then it will be reconverted back to bool here.

If so, I'd prefer we avoid the conversion gymnastics. Can we just change `F._canonical_mask()` somehow to do it right in the first place?

Reply:
Joel, a few considerations, and I'm happy to do this another way:

1 - if we say a mask can be floating point (which we say pervasively everywhere), then we shoulkd accept it here.  
2 - you are right, in this scenario for the fastpath there are two conversions. It's not so easy though...

Considerations:
* we need a very detailed set of conditions to find whether this should be float or bool.  Because when it goes to sdpa, float is the preferred way, otherwise we have a conversion in every layer.
* so, we'd have to duplicate the analysis before canonical mask -- or sink canonical mask further down.  That's feasible (at some logic complexity, like we need a "check_only" boolean, and the path will be more interesting.  But manageable
* switching to float not only avoided the conversions on the sdpa, path, it also allowed us to handle mask logic more easily.  Because masks must be folded at various points, and that uses to check wthe data type and eiother add the masks (for float masks) or do logicals (for bool masks).  So we have to add the bool mask merge logic back everywhere.
* double conversion bool-float-bool happens only once on the way to Nested Tensor creation (once for all layers), bool float conversion if a bool is passed to sdpa happens in every layer.  
* we are talking about refactoring the fastpath to take advantage. of PT2 compile (since the current Nested Tensors are not supproted by `torch.compile()`.  So just wanted to switch off annoying warnings reported by our users (and the warnings are frankly incorrect, because even if they pass in bool we convert it to float and warn the users to please pass in bool.  Ummm...)  and improve the general situation (such as allowing float masks in nested tensor creation).

This is not a perfect solution, I did feel it made things notably better.  In my mind a perfect solution is more expensive, and actually not that perfect, and not again, the refactor just screamed make it better with limited engineering.  Doing nothing feels like less than this.